src/include/duckdb/function/udf_function.hpp

```diff
@@ -0,0 +1,327 @@
+	    vector<SQLType> arguments;
+	    GetArgumentTypesRecursive<Args...>(arguments);
+
+	    SQLType ret_type = GetArgumentType<TR>();
+
+	    RegisterFunction(name, arguments, ret_type, udf_function, context);
+	}
+
+	static void RegisterFunction(string name, vector<SQLType> args, SQLType ret_type,
+								 scalar_function_t udf_function, ClientContext &context) {
+
+		ScalarFunction scalar_function = ScalarFunction(name, args, ret_type, udf_function);
+		CreateScalarFunctionInfo info(scalar_function);
+		InsertFunctionIntoCatalog(&info, context);
+	}
+
+private:
+	//-------------------------------- Templated functions --------------------------------//
+	template<typename TR, typename... Args>
+	static scalar_function_t CreateUnaryFunction(string name, TR (*udf_func)(Args...)) {
+		assert(sizeof...(Args) == 1);
+		return CreateUnaryFunction<TR, Args...>(name, udf_func);
+	}
+
+	template<typename TR, typename TA>
+	static scalar_function_t CreateUnaryFunction(string name, TR (*udf_func)(TA)) {
+		scalar_function_t udf_function = [=] (DataChunk &input, ExpressionState &state, Vector &result) -> void {
+										UnaryExecutor::Execute<TA, TR>(input.data[0],
+																	   result,
+																	   input.size(),
+																	   udf_func);
+									};
+		return udf_function;
+	}
+
+	template<typename TR, typename... Args>
+	static scalar_function_t CreateBinaryFunction(string name, TR (*udf_func)(Args...)) {
+		assert(sizeof...(Args) == 2);
+		return CreateBinaryFunction<TR, Args...>(name, udf_func);
+	}
+
+	template<typename TR, typename TA, typename TB>
+	static scalar_function_t CreateBinaryFunction(string name, TR (*udf_func)(TA, TB)) {
+		scalar_function_t udf_function = [=] (DataChunk &input, ExpressionState &state, Vector &result) -> void {
+										BinaryExecutor::Execute<TA, TB, TR>(input.data[0],
+																		   	input.data[1],
+																		    result,
+																		    input.size(),
+																		    udf_func);
+									};
+		return udf_function;
+	}
+
+	template<typename TR, typename... Args>
+	static scalar_function_t CreateTernaryFunction(string name, TR (*udf_func)(Args...)) {
+		assert(sizeof...(Args) == 3);
+		return CreateTernaryFunction<TR, Args...>(name, udf_func);
+	}
+
+	template<typename TR, typename TA, typename TB, typename TC>
+	static scalar_function_t CreateTernaryFunction(string name, TR (*udf_func)(TA, TB, TC)) {
+		scalar_function_t udf_function = [=] (DataChunk &input, ExpressionState &state, Vector &result) -> void {
+										TernaryExecutor::Execute<TA, TB, TC, TR>(input.data[0],
+																		   	   	 input.data[1],
+																		   	   	 input.data[2],
+																				 result,
+																				 input.size(),
+																				 udf_func);
+									};
+		return udf_function;
+	}
+
+	static void InsertFunctionIntoCatalog(CreateFunctionInfo *info, ClientContext &context) {
```

Comment:
This transaction management is quite complicated and we would prefer not to have it duplicated all over the place (in case we want to change it later): could you replace this with a call to `void ClientContext::RunFunctionInTransaction`?

This block is equivalent to:
```cpp
context.RunFunctionInTransaction([&]() {
    context.temporary_objects.get()->CreateFunction(context, info);
});
```